src/Generators/Microsoft.Gen.Logging/Emission/Emitter.Method.cs

```diff
@@ -348,43 +389,68 @@ void GenPropertyLoads(LoggingMethod lm, string stateName, out int numReservedUnc
-
-                            var skipNull = member.PotentiallyNull && !member.IsEnumerable
-                                ? $"if ({value} != null) "
-                                : string.Empty;
+                            var classification = MakeClassificationValue(member.ClassificationAttributeTypes);
 
-                            OutLn($"{skipNull}{stateName}.AddTag(\"{propName}\", {value});");
+                            if (member.PotentiallyNull)
+                            {
+                                OutOpenBrace();
+                                OutLn($"var {tmpVarName} = {value};");
+                                OutLn($"if ({tmpVarName} != null)");
+                                OutOpenBrace();
+                                OutLn($"{stateName}.AddClassifiedTag(\"{propName}\", {tmpVarName}, {classification});");
+                                OutCloseBrace();
+                                OutCloseBrace();
+                                OutLn();
+                            }
+                            else
+                            {
+                                OutLn($"{stateName}.AddClassifiedTag(\"{propName}\", {value}, {classification});");
+                            }
                         }
                         else
                         {
-                            var propName = PropertyChainToString(propertyChain, member, "_", omitParameterName: p.OmitParameterName);
+                            var propName = PropertyChainToString(propertyChain, member, "_", omitReferenceName: p.OmitReferenceName);
                             var accessExpression = PropertyChainToString(propertyChain, member, "?.", nonNullSeparator: ".");
 
-                            var value = ConvertPropertyToString(member, accessExpression);
-                            var classification = $"_{EncodeTypeName(member.ClassificationAttributeType!)}_Classification";
+                            var ts = ShouldStringifyProperty(member)
+                                ? ConvertPropertyToString(member, accessExpression)
+                                : accessExpression;
 
-                            var skipNull = member.PotentiallyNull
-                                ? $"if ({value} != null) "
-                                : string.Empty;
+                            var value = member.IsEnumerable
+                                ? $"{LoggerMessageHelperType}.Stringify({accessExpression})"
+                                : ts;
 
-                            OutLn($"{skipNull}{stateName}.AddClassifiedTag(\"{propName}\", {value}, {classification});");
+                            if (member.PotentiallyNull && !member.IsEnumerable)
+                            {
+                                OutOpenBrace();
+                                OutLn($"var {tmpVarName} = {value};");
```

Comment:
Shouldn't we check whether an `IEnumerable` member is `null` before strigifying it?

Reply:
Indeed, nice catch!